hal/inc/ble_hal.h

```diff
@@ -15,190 +15,747 @@
+ * @returns     The current TX power.
+ */
+int8_t ble_gap_get_tx_power(void);
+
+/**
+ * Set the BLE advertising parameters. Changing the advertising parameter during
+ * advertising will restart advertising using the new parameters.
+ *
+ * @param[in]   adv_params  Pointer to the advertising parameters to be set.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_advertising_parameters(const hal_ble_adv_params_t* adv_params, void* reserved);
+
+
+/**
+ * Get the current BLE advertising parameters.
+ *
+ * @param[in,out]   adv_params  Pointer to the advertising parameters to be filled.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_get_advertising_parameters(hal_ble_adv_params_t* adv_params, void* reserved);
+
+/**
+ * Set the BLE advertising data. It will update the advertising data immediately if success.
+ * Changing the advertising data during advertising will restart advertising using the
+ * previous advertising parameters and the new advertising data.
+ *
+ * @param[in]   buf     Pointer to the advertising data to be set.
+ * @param[in]   len     Length of the advertising data.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_advertising_data(const uint8_t* buf, uint16_t len, void* reserved);
+
+/**
+ * Get the current BLE advertising data.
+ *
+ * @param[in,out]   buf     Pointer to the advertising data to be filled.
+ * @param[in]       len     Length of the given buffer.
+ *
+ * @returns     The length of copied data.
+ */
+size_t ble_gap_get_advertising_data(uint8_t* buf, uint16_t len, void* reserved);
+
+/**
+ * Set the BLE scan response data.
+ *
+ * @param[in]   buf     Pointer to the scan response data to be set.
+ * @param[in]   len     Length of the scan response data.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_scan_response_data(const uint8_t* buf, uint16_t len, void* reserved);
+
+/**
+ * Get the current BLE scan response data.
+ *
+ * @param[in,out]   buf     Pointer to the scan response data to be filled.
+ * @param[in]       len     Length of the given buffer.
+ *
+ * @returns     The length of copied data.
+ */
+size_t ble_gap_get_scan_response_data(uint8_t* buf, uint16_t len, void* reserved);
+
+/**
+ * Start BLE advertising.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_start_advertising(void* reserved);
+
+/**
+ * Stop BLE advertising.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_stop_advertising(void);
+
+/**
+ * Check if BLE is advertising.
+ *
+ * @returns     true if it is advertising, otherwise, false.
+ */
+bool ble_gap_is_advertising(void);
+
+/**
+ * Set the BLE scanning parameters.
+ *
+ * @param[in]   scan_params Pointer to the scanning parameters to be set.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_scan_parameters(const hal_ble_scan_params_t* scan_params, void* reserved);
```

Comment:
Consider removing `reserved` argument as `hal_ble_scan_params_t` is already versioned.